src/vs/workbench/browser/parts/editor/tabsTitleControl.ts

```diff
@@ -1292,6 +1292,22 @@ registerThemingParticipant((theme: ITheme, collector: ICssStyleCollector) => {
 		if (tabActiveBackground && adjustedTabBackground && adjustedTabDragBackground) {
 			const adjustedColor = tabActiveBackground.flatten(adjustedTabBackground);
 			const adjustedColorDrag = tabActiveBackground.flatten(adjustedTabDragBackground);
+			collector.addRule(`
```

Comment:
@dmolin I feel like there is room for optimization here because in 99% of the times, the 2 colors will be the same (tab active background, tab unfocused active background). Can we rewrite this such as we use the old CSS rule we used to have if both colors are the same and only split them up otherwise?

Reply:
uhm.. I'm not sure here.
I actually followed the way the code is currently laid out (the same exact patter is already used for `tabHoverBackground` and `tabUnfocusedHoverBackground`).

Moreover, now the 2 CSS selectors are substantially different (the one for the active background explicitely uses `group-container.active`, the other doesn't). So, they would never share the current CSS (since one editor group is active, the other is not). There are ways to improve by generating the rule via composition (a prefix string that is different for the 2 rules plus a common one extracted into a variable) but the things is that this is not the approach taken for the other rules in this code block; I tried to maintain style consistency here, so I laid out the rule in the same way the others have already been written

Reply:
@bpasero what I mean is, to make it work without code redundancies I could do this:

```
		const makeTabActiveBackgroundSelector = (color: Color, colorDrag: Color, isContainerActive = false) => `
				.monaco-workbench .part.editor > .content:not(.dragged-over) .editor-group-container${isContainerActive ? ".active" : ""} > .title .tabs-container > .tab.sizing-shrink.active:not(.dragged) > .tab-label::after {
					background: linear-gradient(to left, ${color}, transparent);
				}

				.monaco-workbench .part.editor > .content.dragged-over .editor-group-container${isContainerActive ? ".active" : ""} > .title .tabs-container > .tab.sizing-shrink.active:not(.dragged) > .tab-label::after {
					background: linear-gradient(to left, ${colorDrag}, transparent);
				}
		`;

		// Adjust gradient for active tab background
		const tabActiveBackground = theme.getColor(TAB_ACTIVE_BACKGROUND);
		if (tabActiveBackground && adjustedTabBackground && adjustedTabDragBackground) {
			const adjustedColor = tabActiveBackground.flatten(adjustedTabBackground);
			const adjustedColorDrag = tabActiveBackground.flatten(adjustedTabDragBackground);
			collector.addRule(makeTabActiveBackgroundSelector(adjustedColor, adjustedColorDrag, true));
		}

		// Adjust gradient for unfocused active tab background
		const tabUnfocusedActiveBackground = theme.getColor(TAB_UNFOCUSED_ACTIVE_BACKGROUND);
		if (tabUnfocusedActiveBackground && adjustedTabBackground && adjustedTabDragBackground) {
			const adjustedColor = tabUnfocusedActiveBackground.flatten(adjustedTabBackground);
			const adjustedColorDrag = tabUnfocusedActiveBackground.flatten(adjustedTabDragBackground);
			collector.addRule(makeTabActiveBackgroundSelector(adjustedColor, adjustedColorDrag));
		}
```

But that would be quite a different way to proceed compared to what has been already done for the aforementioned `tabHoverBackground` and `tabUnfocusedHoverBackground` (unless we agree to apply the same change there too, for the sake of consistency)

Let me know what your thoughts are about this

Reply:
@dmolin yeah I think even the other rules could be tweaked. after all less CSS rules is better and faster if possible.

My idea was the following: if both colors are defined and identical you can simply generate a rule `.monaco-workbench .part.editor > .content:not(.dragged-over) .editor-group-container >` and that will work for both active and inactive groups.

Reply:
@bpasero This could implement the "let's define the CSS rule only once" approach you were referring to:

I'm not a super fan of nested `if` but it's the easiest way to avoid too long ifs statements in the following blocks.
The idea, as you suggested, is that if the 2 colors are defined and equals we apply a common rule, otherwise we define 2 distinct specific rules.  Regarding your comment below related to the 2 colors not being customized, I do believe the common rule still applies (so, no reason to add a `:not(.active)` pseudo selector to the rule), since the 2 variable would always have a value and then would fall within the "common" rule.

```
		// Adjust gradient for active tab background (focused and unfocused editor groups)
		const makeTabActiveBackgroundRule = (color: Color, colorDrag: Color, hasFocus = false) => `
				.monaco-workbench .part.editor > .content:not(.dragged-over) .editor-group-container${hasFocus ? ".active" : ""} > .title .tabs-container > .tab.sizing-shrink.active:not(.dragged) > .tab-label::after {
					background: linear-gradient(to left, ${color}, transparent);
				}

				.monaco-workbench .part.editor > .content.dragged-over .editor-group-container${hasFocus ? ".active" : ""} > .title .tabs-container > .tab.sizing-shrink.active:not(.dragged) > .tab-label::after {
					background: linear-gradient(to left, ${colorDrag}, transparent);
				}
		`;
		const tabActiveBackground = theme.getColor(TAB_ACTIVE_BACKGROUND);
		const tabUnfocusedActiveBackground = theme.getColor(TAB_UNFOCUSED_ACTIVE_BACKGROUND);

		// if both colors (focused and unfocused) are equal, a common rule applies to both, otherwise we apply indivitual rules for
		// the .active and inactive editor groups
		if (tabActiveBackground && tabUnfocusedActiveBackground && adjustedTabBackground && adjustedTabDragBackground
			&& tabActiveBackground.equals(tabUnfocusedActiveBackground)) {
			const adjustedColor = tabActiveBackground.flatten(adjustedTabBackground);
			const adjustedColorDrag = tabActiveBackground.flatten(adjustedTabDragBackground);
			collector.addRule(makeTabActiveBackgroundRule(adjustedColor, adjustedColorDrag));
		} else {
			if (tabActiveBackground && adjustedTabBackground && adjustedTabDragBackground) {
				const adjustedColor = tabActiveBackground.flatten(adjustedTabBackground);
				const adjustedColorDrag = tabActiveBackground.flatten(adjustedTabDragBackground);
				collector.addRule(makeTabActiveBackgroundRule(adjustedColor, adjustedColorDrag, true))
			}

			// Adjust gradient for unfocused active tab background
			if (tabUnfocusedActiveBackground && adjustedTabBackground && adjustedTabDragBackground) {
				const adjustedColor = tabUnfocusedActiveBackground.flatten(adjustedTabBackground);
				const adjustedColorDrag = tabUnfocusedActiveBackground.flatten(adjustedTabDragBackground);
				collector.addRule(makeTabActiveBackgroundRule(adjustedColor, adjustedColorDrag))
			}
		}
```

Reply:
@bpasero I'm not sure that defining a common rule, though, is the best way to go... if the 2 colors are equal (that can happen if they are not customized but ALSO if only the active one is in the settings, since the unfocused one by default gets the same value as the active one), then you would get the "common" rule applied; since that rule does not enforce the `.active` part of the selector, the tab would end up being overridden by the rule that normally applies to "inactive" tabs!
(see image, where you can see that the :after pseudo selector for the active tab as been "taken over" by the rule that normally applies to inactive tabs):

![common-rule](https://user-images.githubusercontent.com/861513/58384011-c7007100-7fd4-11e9-928c-ef68dc8857ec.png)

So, I still believe that having the 2 conditions as before is the best way to go, maybe with the "make...Rule" to avoid writing twice (most of) the same code, like so:

```
		// Adjust gradient for active tab background (focused and unfocused editor groups)
		const makeTabActiveBackgroundRule = (color: Color, colorDrag: Color, hasFocus = false) => `
				.monaco-workbench .part.editor > .content:not(.dragged-over) .editor-group-container${hasFocus ? ".active" : ":not(.active)"} > .title .tabs-container > .tab.sizing-shrink.active:not(.dragged) > .tab-label::after {
					background: linear-gradient(to left, ${color}, transparent);
				}

				.monaco-workbench .part.editor > .content.dragged-over .editor-group-container${hasFocus ? ".active" : ":not(.active)"} > .title .tabs-container > .tab.sizing-shrink.active:not(.dragged) > .tab-label::after {
					background: linear-gradient(to left, ${colorDrag}, transparent);
				}
		`;
		
		// Adjust gradient for unfocused active tab background
		const tabActiveBackground = theme.getColor(TAB_ACTIVE_BACKGROUND);
		if (tabActiveBackground && adjustedTabBackground && adjustedTabDragBackground) {
			const adjustedColor = tabActiveBackground.flatten(adjustedTabBackground);
			const adjustedColorDrag = tabActiveBackground.flatten(adjustedTabDragBackground);
			collector.addRule(makeTabActiveBackgroundRule(adjustedColor, adjustedColorDrag, true))
		}

		// Adjust gradient for unfocused active tab background
		const tabUnfocusedActiveBackground = theme.getColor(TAB_UNFOCUSED_ACTIVE_BACKGROUND);
		if (tabUnfocusedActiveBackground && adjustedTabBackground && adjustedTabDragBackground) {
			const adjustedColor = tabUnfocusedActiveBackground.flatten(adjustedTabBackground);
			const adjustedColorDrag = tabUnfocusedActiveBackground.flatten(adjustedTabDragBackground);
			collector.addRule(makeTabActiveBackgroundRule(adjustedColor, adjustedColorDrag))
		}
```

Reply:
Yeah good point.

Reply:
one question @bpasero : the pre-commit hook doesn't allow me to commit because of this `.editor-group-container${hasFocus ? ".active" : ""}`
It says that there are "unexternalized strings"... where am I supposed to put those constants?

Reply:
@dmolin simply use single quotes instead of double quotes 👍

Reply:
I pushed the latest commit, essentially with the solution mentioned in my last comment (applied also to the `tabHoverBackground` rule. See f763adc